File: CodinGame/Puzzles/Hard/Roller Coaster/RollerCoaster.js
let [ places, rides, groupsCount ] = readline().split(' ').map(Number);
let groups = [], earnings = 0, idx = 0, totalPeople = 0;

for (let i = 0; i < groupsCount; i++) {
    const people = parseInt(readline());
    totalPeople += people;
    groups.push(people);
}

if (totalPeople <= places) {
    return console.log(totalPeople * rides);
}

while (rides--) {
    let availablePlaces = places;

    while (availablePlaces >= groups[idx]) {        
        earnings += groups[idx];
        availablePlaces -= groups[idx];
        idx = idx === groupsCount - 1 ? 0 : idx + 1;
    }

}

console.log(earnings);

File: CodinGame/Puzzles/Hard/Roller Coaster/RollerCoaster.clj
(ns Solution
  (:require [clojure.string :as str])
  (:gen-class))

(defn output [msg] (println msg) (flush))
(defn debug [msg] (binding [*out* *err*] (println msg) (flush)))

(defn -main [& args]
  (let [[places rides groupsCount] (map #(Integer/parseInt %) (filter #(not-empty %) (str/split (read-line) #" ")))]

    (def groups [])
    (def earnings 0)
    (def idx 0)
    (def totalPeople 0)

    (dotimes [i groupsCount]
        (let [people (Integer/parseInt (read-line))]
            (def totalPeople (+ totalPeople people))
            (def groups (conj groups people))
        )
    )

    (if (<= totalPeople places)
        (output (* totalPeople rides))
        (do
            (dotimes [r rides]
                (def availablePlaces places)

                (loop []
                    (if (>= availablePlaces (get groups idx))
                        (do
                            (def earnings (+ earnings (get groups idx)))
                            (def availablePlaces (- availablePlaces (get groups idx)))
                            (if (= idx (- groupsCount 1))
                            (def idx 0)
                            (def idx (+ idx 1)))
                            (recur)
                        )
                    )
                )
            )

            (output earnings)
        )
    )
  )
)